patterns/self_encapsulation.md

```diff
@@ -0,0 +1,655 @@
+
+### Problem:
+
+Typically, we want subclasses to inherit *behavior*. How do we define classes so that behavior is reliable?
+
+### Constraints:
+
+In a collaborating system of objects, the essence of an object is its responsibilities. Other objects rely on it to fulfill these responsibilities. How it fulfills them is relatively unimportant. Even less important is an object's data structure (that is, the number of slots*, the slots' names, and the type of object that is expected to live in those slots). First and foremost, the essence of an object is its responsibilities.
+
+An object fulfills its responsibilities by passing messages*. Programmers usually find it a lot less time-consuming to define message names (protocols) than to define the functional details of how an object responds to them. (Note: I refer here to the time necessary to define these names for the rest of the system. Don't confuse that with the time needed to determine high-quality message names, which should always be done with care.)
+
+Additionally, the object's data structure often plays an important role in how an object responds to the messages it receives. An object cannot do much without eventually manipulating some sort of data. Implementing an object's data structure forces the programmer to make certain trade-offs between memory space and speed. Considering those trade-offs can distract a programmer from a more serious implication of their decisions: **every data structure implemented in a base class affects its subclass.**
+
+Subclasses automatically inherit the data structure of their base class. So, any decisions about data structure are virtually irreversible by subclasses without overriding every base class method that refers to the data structure. Therefore, each time you add detail to the data structure, some implementation options are lost and others imposed on the class and its subclasses. And, the amount of time it takes to refactor anything involving the data structure of a base class is quickly multiplied by the number subclasses affected by the change.
+
+* Slots could refer to storage slots of any kind (memory, database, etc.)
+* Passing messages usually, but not always, refers to a method call. Conceptually, any time you ask an object to perform some action or respond to you in any way, you pass it a message and expect a certain response.
+
+### Solution:
+
+The pain of dealing with this scenario make one desperately want to decouple a class' data structure from the list of responsibilities it is duty-bound to fulfill. The first step toward that goal is to define what's on the object's list of responsibilities.
+
+**Therefore, when creating a new class, the first thing you should do is define its public protocol.** Specify its behavior without thinking about its data structure (such as instance variables, class variables, and so on). Defining your object by its public methods, not by its properties.
```

Comment:
This, of course, is consistent with defining a class in a test-driven fashion.  Assuming one is writing tests first, the tests would define public methods by their method signature such as: